node/beacon-chain-handler/src/importer.rs

```diff
@@ -1,55 +1,64 @@
 //! BeaconBlockImporter consumes blocks that we received from other peers and adds them to the
 //! chain.
-use beacon::types::{BeaconBlock, BeaconBlockBody};
-use chain::BlockChain;
-use node_runtime::{ApplyState, Runtime};
-use primitives::traits::{Block, Header};
-use primitives::types::BlockId;
-use std::sync::Arc;
 use std::cell::RefCell;
-use storage::StateDb;
-use parking_lot::RwLock;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use futures::{Future, future, Stream};
 use futures::sync::mpsc::Receiver;
-use futures::{Future, Stream, future};
-use std::collections::HashSet;
+use parking_lot::RwLock;
 
+use beacon::types::{BeaconBlock, BeaconBlockChain};
+use chain::Block;
+use node_runtime::{ApplyState, Runtime};
+use primitives::hash::CryptoHash;
+use primitives::types::BlockId;
+use shard::{ShardBlock, ShardBlockChain};
+use storage::StateDb;
 
-pub fn create_beacon_block_importer_task(
-    beacon_chain: Arc<BlockChain<BeaconBlock>>,
+pub fn create_block_importer_task(
+    beacon_chain: Arc<BeaconBlockChain>,
+    shard_chain: Arc<ShardBlockChain>,
     runtime: Arc<RwLock<Runtime>>,
     state_db: Arc<StateDb>,
     receiver: Receiver<BeaconBlock>
 ) -> impl Future<Item = (), Error = ()> {
-    let beacon_block_importer = RefCell::new(BeaconBlockImporter::new(
+    let beacon_block_importer = RefCell::new(BlockImporter::new(
         beacon_chain,
+        shard_chain,
         runtime,
         state_db,
     ));
     receiver.fold(beacon_block_importer, |beacon_block_importer, body| {
-        beacon_block_importer.borrow_mut().import_block(body);
+        beacon_block_importer.borrow_mut().import_beacon_block(body);
         future::ok(beacon_block_importer)
     }).and_then(|_| Ok(()))
 }
 
-pub struct BeaconBlockImporter {
-    beacon_chain: Arc<BlockChain<BeaconBlock>>,
+pub struct BlockImporter {
+    beacon_chain: Arc<BeaconBlockChain>,
+    shard_chain: Arc<ShardBlockChain>,
     runtime: Arc<RwLock<Runtime>>,
     state_db: Arc<StateDb>,
     /// Stores blocks that cannot be added yet.
-    pending_blocks: HashSet<BeaconBlock>,
+    pending_beacon_blocks: HashMap<CryptoHash, BeaconBlock>,
```

Comment:
Now that Block structs do not recompute their hashes we implement `Borrow` trait for `BeaconBlock` and `ShardBlock` that would return the reference to the CryptoHash. That in turn would allow having `HashSet<BeaconBlock>` and `HashSet<ShardBlock>` instead of the HashMaps.

Reply:
We need a hash map for shard_blocks, and it kinda weird to use hash set for beacon_blocks.
Also for a O(1) algo here we will need hashmap. Filed as #177

Reply:
`HashSet<BeaconBlock>`, where `BeaconBlock` implements `Borrow` hash exactly the same functionality and algorithmic complexity as `HashMap<CryptoHash, BeaconBlock>` but makes the code cleaner.

Reply:
Can you look up by hash in hashset?
O(1) in #177 is achieved by having hash of the parent -> to block.

Reply:
You can lookup by hash in hashset in O(1) if the item implements the borrow trait that returns the reference to the type that we add in `hash()` of the Hash trait.
See example: https://github.com/nearprotocol/nearcore/blob/master/core/txflow/src/dag/message/mod.rs#L104

Unfortunately that does not address #177 but merely makes the code cleaner. It does not address #177 because Borrow would need to return the parent hash and not its own hash which is prohibited.

Reply:
Btw, this wasn't resolved. Can you look into this?

Reply:
Asked @bowenwang1996  to take a look.

Reply:
Can we have a separate issue for this?

Reply:
Done. https://github.com/nearprotocol/nearcore/issues/189